Code/Framework/AzCore/AzCore/Debug/Trace.h

```diff
@@ -39,14 +37,122 @@ namespace AZ
             None
         };
 
-        class Trace
+        class ITrace
         {
         public:
-            static Trace& Instance()    { return g_tracer; }
+            O3DEKERNEL_API ITrace()
+            {
+                tracer = this;
+            }
+            O3DEKERNEL_API virtual ~ITrace()
+            {
+                tracer = nullptr;
+            }
+            ITrace(const ITrace&) = delete;
+            O3DEKERNEL_API ITrace(ITrace&&) = default;
+            ITrace& operator=(const ITrace&) = delete;
+            O3DEKERNEL_API ITrace& operator=(ITrace&&) = default;
+
+            O3DEKERNEL_API static ITrace& Instance()
+            {
+                if (!tracer)
+                {
+                    static ITrace defaultTracer;
+                }
+                return *tracer;
+            }
+
+            O3DEKERNEL_API static void SetInstance(ITrace* newTracer)
+            {
+                tracer = newTracer;
+            }
```

Comment:
This is some strange logic here.
Because if the ITrace constructor is setting the static instance for the module to *this, then SetInstance can only be used to set an `ITrace` instances that was previously assigned to the `tracer` global to be assigned again to it.

Reply:
The `SetInstance` method is a leftover from a previous design, and it isn't necessary anymore. I'll remove it.